frameworks/modular/module.h

```diff
@@ -0,0 +1,296 @@
+		odp_ ## name ## _subsystem_constructor(void)
+
+#define odp_subsystem_lock(access, name)			\
+	odp_rwlock_ ## access ## _lock(&__subsystem(name).lock)
+
+#define odp_subsystem_unlock(access, name)			\
+	odp_rwlock_ ## access ## _unlock(&__subsystem(name).lock)
+
+/* Base class to all inherited subsystem module classes */
+struct odp_module_base {
+	struct list_node list;
+	const char *name;
+	void *handler; /* DSO */
+	int (*init_local)(void);
+	int (*term_local)(void);
+	int (*init_global)(void);
+	int (*term_global)(void);
+};
+
+/* It is required to define subsystem module class with the
+ * base class as its 1st member and named as "base", and also
+ * use ODP_MODULE_CLASS(subsystem) to create the association
+ * between module class name and subsystem name, like:
+ *
+ * typedef ODP_MODULE_CLASS(subsystem) {
+ *	odp_module_base_t base;
+ *	...new members...
+ * } new_module_t; // Here pick the name you like freely
+ *
+ * It also supports forward declaration like:
+ *
+ * // Forward declaration
+ * typedef ODP_MODULE_CLASS(subsystem) new_module_t;
+ * // Laterly comes the definition
+ * ODP_MODULE_CLASS(subsystem) {
+ *	odp_module_base_t base;
+ *	...new members...
+ * }
+ *
+ * Then in preprocessor macros when we have the subsystem name
+ * we can recover the module class type information, like:
+ *
+ * #define MACRO(subsystem)
+ * do {
+ *	ODP_MODULE_CLASS(subsystem) *mod = NULL;
+ *	odp_subsystem_foreach_module(subsystem, mod) {
+ *		mod->xxx; // access the module class
+ *	}
+ * } while(0)
+ */
+#define ODP_MODULE_CLASS(subsystem) struct odp_ ## subsystem ## _module
+
+/* Below macros assume that all subsystem module classes have
+ * odp_module_base_t as their 1st member named "base".
+ *
+ * This greatly reduces the complexity for module list iteration
+ * and module pointer recovery from its list_node member by a forced
+ * type conversion instead of complex calls to container_of() etc.
+ */
+#define __force_cast(module, node)				\
+	((typeof(module))((void *)(node)))
+
+#define __foreach_module(pos, head)				\
+	for (pos = __force_cast(pos, (head)->node.next);	\
+	     pos != __force_cast(pos, head);			\
+	     pos = __force_cast(pos, (pos)->base.list.next))
+
+#define __foreach_module_safe(pos, n, head)			\
+	for (pos = __force_cast(pos, (head)->node.next),	\
+	     n = __force_cast(pos, (pos)->base.list.next);	\
+	     pos != __force_cast(pos, head);			\
+	     pos = n, n = __force_cast(next, (next)->base.list.next))
+
+#define odp_subsystem_active_module(name, mod)			\
+	__force_cast(mod, __subsystem(name).active)
+
+#define odp_subsystem_foreach_module(name, mod)			\
+	__foreach_module(mod, &__subsystem(name).modules)
+
+#define odp_subsystem_foreach_module_safe(name, mod, next)	\
+	__foreach_module_safe(mod, next, &__subsystem(name).modules)
+
+#define odp_module_constructor(mod) list_node_init(&(mod)->base.list)
+
+/* Module constructors should be late than subsystem constructors,
```

Comment:
Will change to "should be later"

All module constructors should be invoked after subsystem constructors finished?